pkg/controller/installation/core_controller.go

```diff
@@ -1234,6 +1234,12 @@ func (r *ReconcileInstallation) Reconcile(ctx context.Context, request reconcile
 		reqLogger.V(1).Info("Unable to determine MTU - no explicit config, and /var/lib/calico is not mounted")
 	}
 
+	// Check whether the calicoWindowsUpgrader is degraded. If so, requeue
+	// a reconcile.
+	if r.calicoWindowsUpgrader.IsDegraded() {
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
```

Comment:
I'm having a bit of trouble remembering how exactly it's meant to work, but the typha autoscaler also sets degraded from within this loop... https://github.com/tigera/operator/blob/5ee263960b7b814083f77e60a2e1b31663a31ebe/pkg/controller/installation/core_controller.go#L806

@tmjd should we have all SetDegraded calls occurring within the main controller loop? The typha autoscaler seems to set it in two places - within the autoscaler itself, but also within the core controller after checking `IsDegraded` - one of those seems like a duplicate, unless I'm missing something... maybe it's because we can't guarantee a Reconcile() will trigger, so SetDegraded in this main loop won't necessarily be called? If that's the case, why do we need a SetDegraded in core controller?

Reply:
Also, is there a reason not to put this check right next to the typhaAutoscaler equivalent up at line 804? What's the reason for putting it down here?

Reply:
I put the `r.calicoWindowsUpgrader.IsDegraded()` call later in the core `Reconcile()` to minimize the chance of this order of events:
- Partway through core `Reconcile()`, `r.calicoWindows.IsDegraded()` is called but returns false. core `Reconcile()` continues on.
- `updateWindowsNodes()` is triggered by the calicoWindowsUpgrader's internal ticker. The call results in an error and calicoWindowsUpgrader calls `SetDegraded`
- core `Reconcile()` reaches the end and calls `ClearDegraded`

When the calicoWindowsUpgrader's internal ticker ticks again, if `updateWindowsNodes()` returns an error again then it will `SetDegraded` so that's not so bad maybe, but the core `Reconcile()` won't be triggered and we might need to wait 5 minutes for core to reconcile again.

Perhaps this scenario isn't that bad. calico status will be degraded and resolving will involve updating the Windows nodes which would trigger a core reconcile.

Reply:
I don't know that it matters where the SetDegraded is called.
In the case of typhaAutoscaler, the core Reconcile triggers an autoscale attempt and if there is an error from that specific attempt then that is why SetDegraded is called in the core Reconcile (it probably could have also been called by the autoscaler loop that serviced the autoscale request). Otherwise if the autoscale attempt happens because of the autoscaler ticker then the autoscale loop does SetDegraded.

Reply:
I'm thinking this is possibly a better place for the typha autoscaler since the isDegraded for autoscale is followed up with triggering an attempt to autoscale. It is possible that the changes that were just applied in the Reconcile fixed some issue that was preventing autoscaling before.

Reply:
For Windows upgrading, I don't think that race condition is a concern assuming we have the windows upgrader reconciling frequently enough. Even if the upgrader had no errors until after the `r.calicoWindows.IsDegraded()` was called and the Reconcile called ClearDegraded, when the upgrader runs again it will SetDegraded. So assuming our period is low enough, it won't matter if the ClearDegraded removes the initial SetDegraded that happens.
(Am I remembering right and the windows upgrader is being ran on a lowish periodic timer?)

Reply:
> (Am I remembering right and the windows upgrader is being ran on a lowish periodic timer?)

Yes it's a 10s period.